File: services/mcp/remoteInstallPlan.ts
import { randomBytes } from 'crypto'
import { addMcpConfig } from './config.js'
import {
  McpJsonConfigSchema,
  McpServerConfigSchema,
  type ConfigScope,
  type McpServerConfig,
} from './types.js'

export const MCP_REMOTE_PLAN_TOKEN_TTL_MS = 10 * 60 * 1000

export type McpRemoteWritableScope = Extract<
  ConfigScope,
  'local' | 'user' | 'project'
>

export type McpRemoteInstallPlan = {
  token: string
  createdAt: number
  source: string
  serverName: string
  scope: McpRemoteWritableScope
  config: McpServerConfig
  availableServers: string[]
}

export type McpRemotePlanError =
  | {
      type: 'missing_source'
    }
  | {
      type: 'invalid_scope'
      scope?: string
    }
  | {
      type: 'invalid_source'
      reason: string
    }
  | {
      type: 'multiple_servers'
      availableServers: string[]
    }
  | {
      type: 'missing_server_name'
    }
  | {
      type: 'server_not_found'
      serverName: string
      availableServers: string[]
    }
  | {
      type: 'unknown_token'
      token: string
    }
  | {
      type: 'expired_token'
      token: string
    }
  | {
      type: 'install_failed'
      message: string
    }

export type McpRemoteInstallResult =
  | {
      ok: true
      plan: McpRemoteInstallPlan
    }
  | {
      ok: false
      error: McpRemotePlanError
    }

const planStore = new Map<string, McpRemoteInstallPlan>()

type SelectedServerConfig =
  | {
      ok: true
      serverName: string
      config: McpServerConfig
      availableServers: string[]
    }
  | {
      ok: false
      error: McpRemotePlanError
    }

function pruneExpiredPlans(now = Date.now()): void {
  for (const [token, plan] of planStore.entries()) {
    if (now - plan.createdAt > MCP_REMOTE_PLAN_TOKEN_TTL_MS) {
      planStore.delete(token)
    }
  }
}

function createToken(): string {
  let token = randomBytes(4).toString('hex')
  while (planStore.has(token)) {
    token = randomBytes(4).toString('hex')
  }
  return token
}

function normalizeScope(scope?: string): McpRemoteWritableScope | null {
  if (!scope || scope === 'local') return 'local'
  if (scope === 'user' || scope === 'project') return scope
  return null
}

export async function getMcpRemoteInstallPlan(opts: {
  source?: string
  serverName?: string
  scope?: string
}): Promise<McpRemoteInstallResult> {
  pruneExpiredPlans()
  const source = opts.source?.trim()
  if (!source) {
    return { ok: false, error: { type: 'missing_source' } }
  }

  const scope = normalizeScope(opts.scope)
  if (!scope) {
    return { ok: false, error: { type: 'invalid_scope', scope: opts.scope } }
  }

  const loaded = await loadRemoteJson(source)
  if (loaded.ok === false) {
    return { ok: false, error: { type: 'invalid_source', reason: loaded.reason } }
  }

  const selected = selectServerConfig(loaded.json, opts.serverName)
  if (selected.ok === false) return { ok: false, error: selected.error }

  const token = createToken()
  const plan: McpRemoteInstallPlan = {
    token,
    createdAt: Date.now(),
    source,
    serverName: selected.serverName,
    scope,
    config: selected.config,
    availableServers: selected.availableServers,
  }
  planStore.set(token, plan)
  return { ok: true, plan }
}

export async function executeMcpRemoteInstallPlan(
  token: string,
): Promise<McpRemoteInstallResult> {
  pruneExpiredPlans()
  const plan = planStore.get(token)
  if (!plan) {
    return { ok: false, error: { type: 'unknown_token', token } }
  }

  planStore.delete(token)
  if (Date.now() - plan.createdAt > MCP_REMOTE_PLAN_TOKEN_TTL_MS) {
    return { ok: false, error: { type: 'expired_token', token } }
  }

  try {
    await addMcpConfig(plan.serverName, plan.config, plan.scope)
    return { ok: true, plan }
  } catch (error) {
    return {
      ok: false,
      error: {
        type: 'install_failed',
        message: error instanceof Error ? error.message : String(error),
      },
    }
  }
}

function selectServerConfig(
  json: unknown,
  requestedName?: string,
): SelectedServerConfig {
  const mcpJson = McpJsonConfigSchema().safeParse(json)
  if (mcpJson.success) {
    const availableServers = Object.keys(mcpJson.data.mcpServers)
    const serverName =
      requestedName ?? (availableServers.length === 1 ? availableServers[0] : undefined)
    if (!serverName) {
      return {
        ok: false,
        error: { type: 'multiple_servers', availableServers },
      }
    }
    const config = mcpJson.data.mcpServers[serverName]
    if (!config) {
      return {
        ok: false,
        error: {
          type: 'server_not_found',
          serverName,
          availableServers,
        },
      }
    }
    return {
      ok: true,
      serverName,
      config,
      availableServers,
    }
  }

  if (!requestedName) {
    return { ok: false, error: { type: 'missing_server_name' } }
  }
  const serverConfig = McpServerConfigSchema().safeParse(json)
  if (!serverConfig.success) {
    return {
      ok: false,
      error: {
        type: 'invalid_source',
        reason: serverConfig.error.issues
          .map(issue => `${issue.path.join('.')}: ${issue.message}`)
          .join(', '),
      },
    }
  }
  return {
    ok: true,
    serverName: requestedName,
    config: serverConfig.data,
    availableServers: [requestedName],
  }
}

async function loadRemoteJson(
  source: string,
): Promise<{ ok: true; json: unknown } | { ok: false; reason: string }> {
  let url: URL
  try {
    url = new URL(toFetchableUrl(source))
  } catch {
    return {
      ok: false,
      reason: 'Expected an http(s) URL or a GitHub blob URL to a JSON MCP config.',
    }
  }
  if (url.protocol !== 'https:' && url.protocol !== 'http:') {
    return {
      ok: false,
      reason: 'Only http(s) remote MCP config URLs are supported.',
    }
  }

  const response = await fetch(url, {
    headers: {
      Accept: 'application/json,text/plain;q=0.9,*/*;q=0.1',
      'User-Agent': 'mossen-mcp-installer',
    },
  })
  if (!response.ok) {
    return {
      ok: false,
      reason: `Remote config request failed (${response.status})`,
    }
  }
  try {
    return { ok: true, json: JSON.parse(await response.text()) }
  } catch (error) {
    return {
      ok: false,
      reason: `Remote config is not valid JSON: ${error instanceof Error ? error.message : String(error)}`,
    }
  }
}

function toFetchableUrl(source: string): string {
  const url = new URL(source)
  if (url.hostname !== 'github.com' && url.hostname !== 'www.github.com') {
    return source
  }
  const parts = url.pathname.split('/').filter(Boolean)
  if (parts.length >= 5 && parts[2] === 'blob') {
    const [owner, repo, , ref, ...pathParts] = parts
    return `https://raw.githubusercontent.com/${owner}/${repo}/${ref}/${pathParts.join('/')}`
  }
  return source
}

export function _resetMcpRemotePlanStoreForTesting(): void {
  planStore.clear()
}

File: crates/mossen-agent/src/mcp/remote_install_plan.rs
//! Remote MCP config install plan — fetch JSON from URL and staged install.
//!
//! Translates `services/mcp/remoteInstallPlan.ts`.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use rand::Rng;

use crate::mcp::types::{ConfigScope, McpServerConfig};

pub const MCP_REMOTE_PLAN_TOKEN_TTL_MS: u64 = 10 * 60 * 1000;

#[derive(Debug, Clone)]
pub struct McpRemoteInstallPlan {
    pub token: String,
    pub created_at: u64,
    pub source: String,
    pub server_name: String,
    pub scope: ConfigScope,
    pub config: McpServerConfig,
    pub available_servers: Vec<String>,
}

#[derive(Debug, Clone)]
pub enum McpRemotePlanError {
    MissingSource,
    InvalidScope { scope: Option<String> },
    InvalidSource { reason: String },
    MultipleServers { available_servers: Vec<String> },
    MissingServerName,
    ServerNotFound { server_name: String, available_servers: Vec<String> },
    UnknownToken { token: String },
    ExpiredToken { token: String },
    InstallFailed { message: String },
}

pub enum McpRemoteInstallResult {
    Ok { plan: McpRemoteInstallPlan },
    Err { error: McpRemotePlanError },
}

lazy_static::lazy_static! {
    static ref PLAN_STORE: Mutex<HashMap<String, McpRemoteInstallPlan>> = Mutex::new(HashMap::new());
}

fn now_ms() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_millis() as u64
}

fn prune_expired_plans(store: &mut HashMap<String, McpRemoteInstallPlan>) {
    let now = now_ms();
    store.retain(|_, plan| now - plan.created_at <= MCP_REMOTE_PLAN_TOKEN_TTL_MS);
}

fn create_token(store: &HashMap<String, McpRemoteInstallPlan>) -> String {
    let mut rng = rand::thread_rng();
    loop {
        let token = format!("{:08x}", rng.gen::<u32>());
        if !store.contains_key(&token) {
            return token;
        }
    }
}

fn normalize_scope(scope: Option<&str>) -> Option<ConfigScope> {
    match scope {
        None | Some("local") => Some(ConfigScope::Local),
        Some("user") => Some(ConfigScope::User),
        Some("project") => Some(ConfigScope::Project),
        _ => None,
    }
}

/// Convert GitHub blob URLs to raw content URLs.
fn to_fetchable_url(source: &str) -> String {
    match url::Url::parse(source) {
        Ok(u) => {
            let host = u.host_str().unwrap_or("");
            if host != "github.com" && host != "www.github.com" {
                return source.to_string();
            }
            let parts: Vec<&str> = u.path().split('/').filter(|s| !s.is_empty()).collect();
            if parts.len() >= 5 && parts[2] == "blob" {
                let owner = parts[0];
                let repo = parts[1];
                let ref_name = parts[3];
                let path_parts = &parts[4..];
                return format!(
                    "https://raw.githubusercontent.com/{}/{}/{}/{}",
                    owner, repo, ref_name, path_parts.join("/")
                );
            }
            source.to_string()
        }
        Err(_) => source.to_string(),
    }
}

/// Load remote JSON config from URL.
async fn load_remote_json(source: &str) -> Result<serde_json::Value, String> {
    let fetchable = to_fetchable_url(source);
    let url = match url::Url::parse(&fetchable) {
        Ok(u) => u,
        Err(_) => {
            return Err(
                "Expected an http(s) URL or a GitHub blob URL to a JSON MCP config.".to_string(),
            );
        }
    };

    let scheme = url.scheme();
    if scheme != "https" && scheme != "http" {
        return Err("Only http(s) remote MCP config URLs are supported.".to_string());
    }

    let client = reqwest::Client::new();
    let response = client
        .get(url.as_str())
        .header("Accept", "application/json,text/plain;q=0.9,*/*;q=0.1")
        .header("User-Agent", "mossen-mcp-installer")
        .send()
        .await
        .map_err(|e| format!("Remote config request failed: {}", e))?;

    if !response.status().is_success() {
        return Err(format!(
            "Remote config request failed ({})",
            response.status().as_u16()
        ));
    }

    let text = response
        .text()
        .await
        .map_err(|e| format!("Failed to read response: {}", e))?;

    serde_json::from_str(&text)
        .map_err(|e| format!("Remote config is not valid JSON: {}", e))
}

/// Select a server config from the parsed JSON.
fn select_server_config(
    json: &serde_json::Value,
    requested_name: Option<&str>,
) -> Result<(String, McpServerConfig, Vec<String>), McpRemotePlanError> {
    // Try parsing as McpJsonConfig { mcpServers: { ... } }
    if let Some(obj) = json.as_object() {
        if let Some(servers_val) = obj.get("mcpServers") {
            if let Some(servers) = servers_val.as_object() {
                let available: Vec<String> = servers.keys().cloned().collect();
                let server_name = match requested_name {
                    Some(n) => n.to_string(),
                    None => {
                        if available.len() == 1 {
                            available[0].clone()
                        } else {
                            return Err(McpRemotePlanError::MultipleServers {
                                available_servers: available,
                            });
                        }
                    }
                };

                let config_val = match servers.get(&server_name) {
                    Some(v) => v,
                    None => {
                        return Err(McpRemotePlanError::ServerNotFound {
                            server_name,
                            available_servers: available,
                        });
                    }
                };

                let config: McpServerConfig = serde_json::from_value(config_val.clone())
                    .map_err(|e| McpRemotePlanError::InvalidSource {
                        reason: e.to_string(),
                    })?;

                return Ok((server_name, config, available));
            }
        }
    }

    // Try parsing as a bare McpServerConfig
    let name = match requested_name {
        Some(n) => n.to_string(),
        None => return Err(McpRemotePlanError::MissingServerName),
    };

    let config: McpServerConfig =
        serde_json::from_value(json.clone()).map_err(|e| McpRemotePlanError::InvalidSource {
            reason: e.to_string(),
        })?;

    Ok((name.clone(), config, vec![name]))
}

/// Create a remote install plan.
pub async fn get_mcp_remote_install_plan(
    source: Option<&str>,
    server_name: Option<&str>,
    scope: Option<&str>,
) -> McpRemoteInstallResult {
    {
        let mut store = PLAN_STORE.lock().unwrap();
        prune_expired_plans(&mut store);
    }

    let source = match source.map(|s| s.trim()) {
        Some(s) if !s.is_empty() => s.to_string(),
        _ => {
            return McpRemoteInstallResult::Err {
                error: McpRemotePlanError::MissingSource,
            };
        }
    };

    let resolved_scope = match normalize_scope(scope) {
        Some(s) => s,
        None => {
            return McpRemoteInstallResult::Err {
                error: McpRemotePlanError::InvalidScope {
                    scope: scope.map(|s| s.to_string()),
                },
            };
        }
    };

    let json = match load_remote_json(&source).await {
        Ok(j) => j,
        Err(reason) => {
            return McpRemoteInstallResult::Err {
                error: McpRemotePlanError::InvalidSource { reason },
            };
        }
    };

    let (name, config, available) = match select_server_config(&json, server_name) {
        Ok(v) => v,
        Err(e) => return McpRemoteInstallResult::Err { error: e },
    };

    let mut store = PLAN_STORE.lock().unwrap();
    let token = create_token(&store);
    let plan = McpRemoteInstallPlan {
        token: token.clone(),
        created_at: now_ms(),
        source,
        server_name: name,
        scope: resolved_scope,
        config,
        available_servers: available,
    };
    store.insert(token, plan.clone());
    McpRemoteInstallResult::Ok { plan }
}

/// Execute a previously-created remote install plan.
pub async fn execute_mcp_remote_install_plan(
    token: &str,
    add_mcp_config: &dyn super::builtin_template_plan::AsyncAddMcpConfig,
) -> McpRemoteInstallResult {
    let plan = {
        let mut store = PLAN_STORE.lock().unwrap();
        prune_expired_plans(&mut store);
        store.remove(token)
    };

    let plan = match plan {
        Some(p) => p,
        None => {
            return McpRemoteInstallResult::Err {
                error: McpRemotePlanError::UnknownToken { token: token.to_string() },
            };
        }
    };

    if now_ms() - plan.created_at > MCP_REMOTE_PLAN_TOKEN_TTL_MS {
        return McpRemoteInstallResult::Err {
            error: McpRemotePlanError::ExpiredToken { token: token.to_string() },
        };
    }

    match add_mcp_config.add_config(&plan.server_name, &plan.config, plan.scope).await {
        Ok(()) => McpRemoteInstallResult::Ok { plan },
        Err(e) => McpRemoteInstallResult::Err {
            error: McpRemotePlanError::InstallFailed { message: e.to_string() },
        },
    }
}

/// Reset plan store for testing.
pub fn reset_mcp_remote_plan_store_for_testing() {
    PLAN_STORE.lock().unwrap().clear();
}

/// TS `type McpRemoteWritableScope` — the scopes an MCP remote-install plan
/// is permitted to write into.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum McpRemoteWritableScope {
    Project,
    User,
}
